### PR TITLE
Use masked stack to preserve mask info

### LINKED ISSUE #24545: [Bug]: `matplotlib.pyplot.scatter` does not respect mask rules with `datetime`
### Bug summary

Plotting a numpy masked array against a list of datetime objects, `plt.plot` functions as expected, but `plt.scatter` still displays the masked data. Auto-axis limits *do* respect the masked data though.

### Code for reproduction

```python
import matplotlib.pyplot as plt
import numpy.ma as ma
from datetime import datetime

# Create a list of datetime objects
time_unmasked = [datetime(2022, 11, 24, 4, 49, 52),
                 datetime(2022, 11, 24, 4, 49, 53),
                 datetime(2022, 11, 24, 4, 49, 54),
                 datetime(2022, 11, 24, 4, 49, 55),
                 datetime(2022, 11, 24, 4, 49, 56),
                 datetime(2022, 11, 24, 4, 49, 57),
                 datetime(2022, 11, 24, 4, 49, 58)]

# Create a masked array with three masked variables, one datapoint far from others
data_masked = ma.array([15,14,50,15,15,15,16], mask=[0,0,1,1,1,0,0])

# Line plot (expected behaviour)
plt.plot(time_unmasked, data_masked)
plt.show()

# Scatter plot (plots masked points, but ignores masked values when auto-setting axis limits)
plt.scatter(time_unmasked, data_masked)
plt.show()
```

### Actual outcome

`plot` behaves as expected.

[![enter image description here][1]][1]

`scatter` unexpectedly draws the masked datapoints, but axes limits are still respecting the masked array as the masked `y=50` value is not visible.
[![enter image description here][2]][2]

  [1]: https://i.stack.imgur.com/wk1FI.png
  [2]: https://i.stack.imgur.com/6Y6S1.png

### Expected outcome

Scatter should not plot the masked data even though all timestamps are unmasked.
[![enter image description here][3]][3]

  [3]: https://i.stack.imgur.com/necjd.png

### Additional information

It seems as though there is a check within `matplotlib.pyplot.scatter` that is not catching. I'd expect that logic to check that if either `x` or `y` were masked, to not plot that point. The datetime object is able to slip through. I've also tried putting `time_unmasked` within a numpy masked array with the same mask as `masked_data` (so both `x` and `y` have the same mask), but the same unexpected behaviour occurs.

### Operating system

MacOS 11.3.1 (20E241)

### Matplotlib Version

3.5.3

### Matplotlib Backend

module://ipykernel.pylab.backend_inline

### Python version

3.8.11

### Jupyter version

5.7.10

### Installation

conda

### PR DESCRIPTION
**Pull request by chahak13**
## PR Summary
As pointed out in #24545 , `scatter` was not masking datetime correctly. This issue would be true for any data with units because of the following lines

https://github.com/matplotlib/matplotlib/blob/cc85fb66da61f85259d7d99f06faa6285f4ed225/lib/matplotlib/collections.py#L323-L325

Whenever `self.have_units()` was `True`, it would use `np.column_stack` instead of `np.ma.column_stack` which would lead to the loss of mask information. Changing that to the masked version resolves #24545 .

## PR Checklist

**Documentation and Tests**
- [x] Has pytest style unit tests (and `pytest` passes)
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] New plotting related features are documented with examples.

**Release Notes**
- [ ] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [ ] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [ ] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

### CODE REVIEW

**Comment by oscargus** [comment]
Can you please add a test (like just copy the one in #24545)?

(This seems to be the correct fix, although my knowledge of masked is very limited...)

**Comment by rcomer** [comment]
Nice detective work @chahak13!  Could you add a test that would have failed without this change?  Maybe adapt the example from #24545 into an [image test](https://matplotlib.org/devdocs/devel/testing.html#writing-an-image-comparison-test).

**Comment by oscargus** [comment]
It may be that this is actually better solved in `set_offsets`:

https://github.com/matplotlib/matplotlib/blob/cc85fb66da61f85259d7d99f06faa6285f4ed225/lib/matplotlib/collections.py#L548-L550

which would then remove the need for an additional unit conversion etc. Although one probably should move the logic of `get_offsets` for `None` values to `set_offsets` and call `set_offsets` from the constructor. (Have not checked all the consequences and that doesn't really have to be part of this PR...)

**Comment by chahak13** [comment]
@oscargus good point. This might actually be a more deep-rooted issue. I checked `Path` and it does take in masked arrays (which it then converts to NaNs) so that should also use masked array functions. Also, as you pointed out, this is probably something that should be set directly in `set_offsets`. Not all functions use the masked versions, maybe it is worthwhile to put in effort to recognize and change those?

**Comment by chahak13** [comment]
I've added a test and baseline images for this for now. This should resolve the current issue but I'll try to dig into the usage of masked data later. Let me know if there's something missing!

**Comment by oscargus** [comment]
Thanks!

Maybe, for the sake of it, add a `@check_figures_equal` test where the same is used for one of the figures and the other masks the x-data instead?

(It may be enough to only have one of the format, e.g., png, but that would require a rebase with force push as the cleanliness test will become unhappy otherwise.)

**Comment by oscargus** [comment]
Oh, and if there is a way to actually get a masked Path with units to test with that would be great, but shouldn't be a show stopper for now.

File: lib/matplotlib/collections.py [hunk 1]
```diff
@@ -319,10 +319,12 @@ def _prepare_points(self):
                 xs, ys = vertices[:, 0], vertices[:, 1]
                 xs = self.convert_xunits(xs)
                 ys = self.convert_yunits(ys)
-                paths.append(mpath.Path(np.column_stack([xs, ys]), path.codes))
+                paths.append(
+                    mpath.Path(np.ma.column_stack([xs, ys]), path.codes)
+                )
```
Comment (rcomer) [inline lib/matplotlib/collections.py:324]: I'm not sure this is necessary, since each `path` will be a `Path` object, and the masking is removed at instantiation of the `Path` object
https://github.com/matplotlib/matplotlib/blob/f8d932e2bdb8a5b9d8ff282e76c74b297ed5f981/lib/matplotlib/path.py#L129
So I think, when you get to here, `xs` and `ys` cannot be masked?

File: lib/matplotlib/tests/test_collections.py [hunk 2]
```diff
@@ -1129,3 +1130,22 @@ def test_set_offset_units():
     off0 = sc.get_offsets()
     sc.set_offsets(list(zip(y, d)))
     np.testing.assert_allclose(off0, sc.get_offsets())
+
+
+@image_comparison(baseline_images=["test_check_masked_offsets"],
+                  remove_text=True, style="mpl20")
```
Comment (rcomer) [inline lib/matplotlib/tests/test_collections.py:1136]:
```suggestion
                  extensions=["png"], remove_text=True, style="mpl20")
```
I agree with @oscargus that we probably only need png format for the test.  I think the other formats are only necessary if you think the behaviour would be different across formats.

Comment (rcomer) [reply lib/matplotlib/tests/test_collections.py:1136]: And now I found the documentation that says limiting to png is preferred: https://matplotlib.org/devdocs/api/testing_api.html#matplotlib.testing.decorators.image_comparison

File: lib/matplotlib/collections.py (hunk 1 above)
Comment (chahak13) [reply lib/matplotlib/collections.py:324]: Ah okay, yeah. That makes sense. Thanks!

File: lib/matplotlib/tests/test_collections.py (hunk 2 above)
Comment (chahak13) [reply lib/matplotlib/tests/test_collections.py:1136]: Got it. I'll rebase.

**Comment by chahak13** [comment]
@oscargus @rcomer, need advice. I was working on the `@check_figures_equal` test and it wasn't working correctly if I masked the datetime array instead of `y`. On further digging, this goes into the maintenance issue that Oscar just opened. 

TLDR: `date2num` converts the masked data array to a normal numpy array which is why the test was failing. What do you suggest I should do? I can't write the test to compare the two cases of masking without this.

----
https://github.com/matplotlib/matplotlib/blob/cc85fb66da61f85259d7d99f06faa6285f4ed225/lib/matplotlib/axis.py#L1731-L1740

The reason the fix that I made worked for this case is that the mask was on a numeric field. For all the non-number types, `_is_natively_supported()` returns `False` and then asks the axis' converter to convert the data. The data contains the mask information until this point, so every converter in the registry should be able to handle the masked array by either using masked array functions or raising a warning about it to ensure that the user doesn't expect the masked information in the plot.

**Comment by oscargus** [comment]
I'd say skip the equal test then for now. It may be a deep rabbit hole...

I'll add a comment to #24733 pointing this out. Probably the right thing is to support masked arrays in `date2num`, but that may lead to other strange things...

**Comment by chahak13** [comment]
That should do it. Let me know if there's anything needed!

**Review by rcomer — APPROVED** [review]
Thanks @chahak13, this looks good to me (assuming CI passes, but I can't see a reason it wouldn't).

**Comment by oscargus** [comment]
Thanks! FYI, it wasn't that tricky to get masked dates to pass: https://github.com/matplotlib/matplotlib/pull/24734

But we can add the image equal test when digging into the other aspects later.

**Review by oscargus — APPROVED** [review]
Agreed! Looks good! Anyone can merge on green.

**Comment by chahak13** [comment]
Yes, it was pretty straight forward! I had the fix on my local machine but just wasn't sure if we wanted to include that in this or would that be out of scope for this particular PR. Thanks though!

**Comment by oscargus** [comment]
Ahh, sorry about that then. Just didn't want you to go down a potential rabbit hole by encouraging it...
